pandas/core/arrays/base.py

```diff
@@ -301,6 +319,64 @@ def unique(self):
         uniques = unique(self.astype(object))
         return self._constructor_from_sequence(uniques)
 
+    def _values_for_factorize(self):
+        """Return an array and missing value suitable for factorization.
+
+        Returns
+        -------
+        values : ndarray
+            An array suitable for factoraization. This should maintain order
+            and be a supported dtype. By default, the extension array is cast
+            to object dtype.
+        na_value : scalar
```

Comment:
OK, here's the meaningful change from my last round of commits. We have the user specify an ndarray of values to be factorized, and the na value *going into* the factorization routine.

Previously, `Categorical.factorize` had to be implemented, since it used integers and `iNaT` for the array and missing value in `_factorize_array`. So from a code reuse and interface flexibility point of view, this is a good thing.

I dislike how this opens up the internals of pandas' factorization algorithms to the public more. Is it publicly documented anywhere that iNaT is the minimum integer? How would an EA author know to do something like what we do for Categorical (and PeriodArray when we get there)?